packages/docusaurus/lib/client/exports/Link.js

```diff
@@ -5,16 +5,57 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import React from 'react';
+import React, {useEffect} from 'react';
 import Perimeter from 'react-perimeter';
 import {NavLink} from 'react-router-dom';
 
 const internalRegex = /^\/(?!\/)/;
 
+// Set up IntersectionObserver
+const handleIntersection = (el, cb) => {
+  const io = new window.IntersectionObserver(entries => {
+    entries.forEach(entry => {
+      if (el === entry.target) {
+        // Check if element is within viewport, remove listener, destroy observer, and run link callback.
+        // MSEdge doesn't currently support isIntersecting, so also test for  an intersectionRatio > 0
+        if (entry.isIntersecting || entry.intersectionRatio > 0) {
+          io.unobserve(el);
+          io.disconnect();
+          cb();
+        }
+      }
+    });
+  });
+  // Add element to the observer
+  io.observe(el);
+};
+
 function Link(props) {
   const {to, href, preloadProximity = 20} = props;
   const targetLink = to || href;
   const isInternal = internalRegex.test(targetLink);
+
+  const IOSupported =
+    typeof window !== 'undefined' && 'IntersectionObserver' in window;
+
+  const handleRef = ref => {
+    if (IOSupported && ref) {
+      // If IO supported and element reference found, setup Observer functionality
+      handleIntersection(ref, () => {
+        if (isInternal) {
+          window.__docusaurus.prefetch(targetLink);
+        }
+      });
+    }
+  };
+
+  useEffect(() => {
+    // If IO is not supported. We prefetch by default
+    if (!IOSupported && isInternal) {
+      window.__docusaurus.prefetch(targetLink);
+    }
+  }, []);
```

Comment:
You probably need to put the variables you are referencing within `useEffect`'s 2nd argument. Here's a 49 min read lol: https://overreacted.io/a-complete-guide-to-useeffect/

Reply:
True ! Should not run this every render. IDK why I removed it lol

Reply:
No, you're not running this on every render now. You're only running it once on mount regardless of whether the link changes.

Reply:
Hmmmm but isn't IOSupported and isInternal and targetLink is never changed.

Reply:
`targetLink` is derived from props and it can change. Just because a component is not unmounted doesn't mean it won't change. Example:

You're on docs/foo and you navigate to docs/bar, if there's no code splitting, the pagination links at the bottom of the page do not get unmounted at all. The props passed into the pagination links simply change because of virtual DOM diffing. People commonly trigger side effects on mount but forgot to do that in `componentDidUpdate` and `useEffect` was created to try to solve that problem.